File: schemas/content-moderation/mongodb/keyword_list_entries.js
// keyword_list_entries: Individual words/phrases in keyword lists.
// See README.md for full design rationale.

const mongoose = require("mongoose");

const keywordListEntriesSchema = new mongoose.Schema(
  {
    list_id: {
      type: mongoose.Schema.Types.ObjectId,
      ref: "KeywordList",
      required: true,
    },
    value: {
      type: String,
      required: true,
    },
    match_type: {
      type: String,
      enum: ["exact", "contains", "regex"],
      required: true,
      default: "exact",
    },
    is_case_sensitive: {
      type: Boolean,
      required: true,
      default: false,
    },
    added_by: {
      type: mongoose.Schema.Types.ObjectId,
      ref: "User",
      required: true,
    },
  },
  {
    timestamps: { createdAt: "created_at", updatedAt: false },
  }
);

// Indexes
keywordListEntriesSchema.index(
  { list_id: 1, value: 1, match_type: 1 },
  { unique: true }
);
module.exports = mongoose.model("KeywordListEntry", keywordListEntriesSchema);

File: schemas/content-moderation/spacetimedb/keyword_list_entries.rs
// keyword_list_entries: Individual words/phrases in keyword lists.
// See README.md for full design rationale.

use spacetimedb::Timestamp;

#[derive(SpacetimeType, Clone)]
pub enum KeywordMatchType {
    Exact,    // type: String
    Contains,
    Regex,
}

#[spacetimedb::table(name = keyword_list_entries, public)]
pub struct KeywordListEntry {
    #[primary_key]
    pub id: String, // UUID

    #[index(btree)]
    pub list_id: String, // FK → keyword_lists.id (cascade delete)

    pub value: String,

    pub match_type: KeywordMatchType,

    pub is_case_sensitive: bool,

    #[index(btree)]
    pub added_by: String, // FK → users.id (restrict delete)

    pub created_at: Timestamp,
}

// Composite unique: (list_id, value, match_type) — enforce in reducer logic.
// SpacetimeDB does not support composite unique constraints in the schema.
